base_layer/core/src/base_node/state_machine_service/states/block_sync.rs

```diff
@@ -438,43 +444,67 @@ async fn request_and_add_blocks<B: BlockchainBackend + 'static>(
     }
     let config = shared.config.block_sync_config;
     for attempt in 0..config.max_add_block_retry_attempts {
-        let (blocks, sync_peer) = request_blocks(shared, sync_peers, block_nums.clone()).await?;
+        let (blocks, sync_peer) = request_blocks(shared, sync_peers, &block_nums).await?;
         if let StateInfo::BlockSync(ref mut info) = shared.info {
             // assuming the numbers are ordered
             info.tip_height = block_nums[block_nums.len() - 1];
         }
         shared.publish_event_info();
+
         for block in blocks {
-            let block_hash = block.hash();
+            let block_height = block.header.height;
+            let block_hash_hex = block.hash().to_hex();
             if let StateInfo::BlockSync(ref mut info) = shared.info {
-                info.local_height = block.header.height;
+                info.local_height = block_height;
             }
 
             shared.publish_event_info();
-            match shared
-                .local_node_interface
-                .submit_block(block.clone(), Broadcast::from(false))
-                .await
-            {
-                Ok(_) => {
+            match async_db::add_block(shared.db.clone(), block.clone()).await {
```

Comment:
I think this is a move in the wrong direction for code maintainability. The code had this in the past, and we moved away from this because there is certain logic that needs to be handled when a block is added. 
Now we just add blocks and its up to the "user" to ensure the logic is handled correctly. As far as I read the code, now I think we already have a bug in the logic. I think we should just have fn submit_block(), and all logic should reside in there. That way we cannot forgot to notify a client, notify them in the correct way etc.

Reply:
Disagree. The handling of the result below is an indication that different parts of the code will handle this event differently. Which parts of the code require handling on a block added? Perhaps those are too tightly coupled.